src/queries.js

```diff
@@ -26,31 +26,25 @@ export function summaryQuery(windowbucket, afkbucket, count) {
   return querystr_to_array(code);
 }
 
-export function windowQuery(
-  windowbucket,
-  afkbucket,
-  appcount,
-  titlecount,
-  filterAFK
-) {
+export function windowQuery(windowbucket, afkbucket, appcount, titlecount, filterAFK, classes) {
   windowbucket = windowbucket.replace('"', '\\"');
   afkbucket = afkbucket.replace('"', '\\"');
   let code =
     `events  = flood(query_bucket("${windowbucket}"));
      not_afk = flood(query_bucket("${afkbucket}"));
      not_afk = filter_keyvals(not_afk, "status", ["not-afk"]);` +
-    (filterAFK ? "events  = filter_period_intersect(events, not_afk);" : "") +
-    `title_events  = merge_events_by_keys(events, ["app", "title"]);
-    title_events  = sort_by_duration(title_events);
-    app_events  = merge_events_by_keys(title_events, ["app"]);
-    app_events  = sort_by_duration(app_events);
+    (filterAFK ? 'events  = filter_period_intersect(events, not_afk);' : '') +
+    `events = classify(events, ${JSON.stringify(classes)});
+    title_events = sort_by_duration(merge_events_by_keys(events, ["app", "title"]));
+    app_events   = sort_by_duration(merge_events_by_keys(title_events, ["app"]));
+    cat_events   = sort_by_duration(merge_events_by_keys(events, ["$category"]));
 
     events = sort_by_timestamp(events);
     app_chunks = chunk_events_by_key(events, "app");
     app_events  = limit_events(app_events, ${appcount});
     title_events  = limit_events(title_events, ${titlecount});
     duration = sum_durations(events);
-    RETURN  = {"app_events": app_events, "title_events": title_events, "app_chunks": app_chunks, "duration": duration};`;
+    RETURN  = {"app_events": app_events, "title_events": title_events, "cat_events": cat_events, "app_chunks": app_chunks, "duration": duration};`;
```

Comment:
As we have learned from earlier when we didn't have proper queries, sending all events to the web-ui and computing it there is not scalable for more than a days worth of data.
You should instead for example utilize the merge_events_by_keys for categories and write some new transform in the server for tags.

Reply:
But I am using `merge_events_by_keys`? I'm basically only sending one event per category in the `cat_events`.